cmd/elasticsearch/cmd/precompute/precompute.go

```diff
@@ -0,0 +1,171 @@
+			return err
+		}
+		return nil
+	},
+	PostRun: func(cmd *cobra.Command, args []string) {
+		logger.Log.Info("Finished 'elasticsearch precompute'")
+	},
+}
+
+// package init defines the flags for the precompute command
+func init() {
+	precomputeCmd.Flags().BoolVar(&updateES, "update", false, "when false, only prints potential updates to stdout instead of touching documents in elasticsearch")
+}
+
+func run(cmd *cobra.Command) error {
+	esClient, err := elasticsearch.NewClient(config.ElasticSearch{
+		Endpoint: cmd.Flag("endpoint").Value.String(),
+		Username: cmd.Flag("user").Value.String(),
+		Password: cmd.Flag("password").Value.String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	pageSize := 10
+	path, payload := BuildScrollQueryInitial("testmachinery-*", pageSize)
+
+	var processedItems int
+	for i := 0;; i++ {
+		esResponse, err := queryAndRecomputeAndStore(esClient, path, payload)
+		if err != nil {
+			return err
+		}
+		processedItems += len(esResponse.Hits.Results)
+		logger.Log.Info(fmt.Sprintf("%d%% processed (%d/%d)", processedItems * 100 / esResponse.Hits.Total.Value, processedItems, esResponse.Hits.Total.Value))
+
+		// once we hit the first page with less than pageSize results, there will be no further page -> done
+		if len(esResponse.Hits.Results) < pageSize {
+			break
+		}
+
+		// on to the next page
+		scrollID := esResponse.ScrollID
+		path, payload = BuildScrollQueryNextPage(scrollID)
+	}
+
+	return nil
+}
+
+// queryAndRecomputeAndStore queries the testmachinery index for tm data, recomputes the pre-calculated fields and if changed, updates the respective data
+func queryAndRecomputeAndStore(esClient elasticsearch.Client, path string, payload io.Reader) (QueryResponse, error) {
+	ctx := context.Background()
+	defer ctx.Done()
+
+	bytes, err := esClient.RequestWithCtx(ctx, http.MethodGet, path, payload)
+	if err != nil {
+		return QueryResponse{}, err
+	}
+
+	var esResponse QueryResponse
+	if err := json.Unmarshal(bytes, &esResponse); err != nil {
+		return QueryResponse{}, err
+	}
+	logger.Log.V(6).Info("Got response from ES", "esResponse", esResponse)
+
+	var modifiedItems []Result
+	for _, result := range esResponse.Hits.Results {
+		var k8sVersion, clusterDomain string
+		if result.StepSummary.Metadata != nil {
+			k8sVersion = result.StepSummary.Metadata.Metadata.KubernetesVersion
+			if k8sVersion == "" && result.StepSummary.Metadata.Annotations != nil {
+				k8sVersion = result.StepSummary.Metadata.Annotations["metadata.testmachinery.gardener.cloud/k8sVersion"]
+			}
+			if k8sVersion == "" && result.StepSummary.Metadata.Annotations != nil {
+				k8sVersion = result.StepSummary.Metadata.Annotations["testrunner.testmachinery.sapcloud.io/k8sVersion"]
+			}
+		}
+		oldPreComputed := result.StepSummary.PreComputed
+		if oldPreComputed != nil {
+			clusterDomain = oldPreComputed.ClusterDomain
+		}
+		newPreComputed := collector.PreComputeTeststepFields(result.StepSummary.Phase, k8sVersion, clusterDomain)
```

Comment:
you dont want to add a cluster domain if the old precomputed is not given.
Doesnt that mean that the cluster is only added if there are already precomputed fields?

Reply:
I struggled a bit to recompute that value in case this field is not already present (I'm not sure which data I can rely on for old datasets, so I don't want to set "some" clusterdomain there in case the information is not present).